engine/render/src/render/buffer.cpp

```diff
@@ -0,0 +1,146 @@
+    }
+
+    HBufferedRenderBuffer NewBufferedRenderBuffer(HRenderContext render_context, RenderBufferType type)
+    {
+        BufferedRenderBuffer* buffer = new BufferedRenderBuffer();
+        memset(buffer, 0, sizeof(BufferedRenderBuffer));
+        RewindBuffer(render_context, buffer);
+        buffer->m_Type = type;
+        return buffer;
+    }
+
+    void DeleteBufferedRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        for (int i = 0; i < buffer->m_Buffers.Size(); ++i)
+        {
+            DeleteRenderBuffer(buffer->m_Buffers[i], buffer->m_Type);
+        }
+        delete buffer;
+    }
+
+    HRenderBuffer AdvanceRenderBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return 0;
+
+        if (buffer->m_Buffers.Empty())
+        {
+            buffer->m_Buffers.OffsetCapacity(1);
+            CreateAndPush(render_context, buffer);
+        }
+
+        if (!render_context->m_MultiBufferingRequired)
+        {
+            buffer->m_BufferIndex = 0;
+            assert(buffer->m_Buffers.Size() == 1);
+            return buffer->m_Buffers.Front();
+        }
+
+        buffer->m_BufferIndex++;
+        if (buffer->m_BufferIndex >= buffer->m_Buffers.Capacity())
+        {
+            buffer->m_Buffers.OffsetCapacity(1);
+            CreateAndPush(render_context, buffer);
+        }
+        return buffer->m_Buffers[buffer->m_BufferIndex];
+    }
+
+    void SetBufferData(HRenderContext render_context, HBufferedRenderBuffer buffer, uint32_t size, void* data, dmGraphics::BufferUsage buffer_usage)
+    {
+        switch(buffer->m_Type)
+        {
+            case RENDER_BUFFER_TYPE_VERTEX_BUFFER:
+            {
+                dmGraphics::HVertexBuffer vbuf = (dmGraphics::HVertexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
+                dmGraphics::SetVertexBufferData(vbuf, size, data, buffer_usage);
+            } break;
+            case RENDER_BUFFER_TYPE_INDEX_BUFFER:
+            {
+                dmGraphics::HIndexBuffer ibuf = (dmGraphics::HIndexBuffer) buffer->m_Buffers[buffer->m_BufferIndex];
+                dmGraphics::SetIndexBufferData(ibuf, size, data, buffer_usage);
+            } break;
+            default:assert(0);
+        }
+    }
+
+    void TrimBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
+
+        uint32_t n = buffer->m_Buffers.Size();
+        uint32_t new_buffer_count = buffer->m_BufferIndex+1;
+
+        if (new_buffer_count == n)
+            return;
+
+        for (int i = 0; i < (n - new_buffer_count); ++i)
+        {
+            PopAndDelete(render_context, buffer);
+        }
+
+        buffer->m_Buffers.SetCapacity(new_buffer_count);
+    }
+
+    void RewindBuffer(HRenderContext render_context, HBufferedRenderBuffer buffer)
+    {
+        if (!buffer)
+            return;
```

Comment:
What is the use case here?
Will the HRenderedBuffer become null in a component, and we still try to render it?
I wonder if it's better to handle it at the component level? E.g. disabling the component world in that case?

Reply:
It's just a shortcut to avoid having to constantly check the buffer for existance in the components since they create the buffers at different times (not necessary at world init, e.g sprites). 

I don't have a strong opinion. Maybe it's better to always create a render buffer when creating a component world? The default single buffer that is created will not allocate any memory anyway so it might as well always be present? 🤔